// @lighthouse/common/dist/scheduling/scheduling.types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StrategyTypes = exports.Unit = exports.IntervalTypes = void 0;
let IntervalTypes;
exports.IntervalTypes = IntervalTypes;

(function (IntervalTypes) {
  IntervalTypes["Occurrence"] = "occurrence";
  IntervalTypes["Service"] = "service";
})(IntervalTypes || (exports.IntervalTypes = IntervalTypes = {}));

let Unit;
exports.Unit = Unit;

(function (Unit) {
  Unit["Millisecond"] = "millisecond";
  Unit["Second"] = "second";
  Unit["Minute"] = "minute";
  Unit["Hour"] = "hour";
  Unit["Day"] = "day";
  Unit["Week"] = "isoWeek";
  Unit["Month"] = "month";
  Unit["Year"] = "year";
})(Unit || (exports.Unit = Unit = {}));

let StrategyTypes;
exports.StrategyTypes = StrategyTypes;

(function (StrategyTypes) {
  StrategyTypes["ExactDateOfMonth"] = "exactDateOfMonth";
  StrategyTypes["LastDayOfMonth"] = "lastDayOfMonth";
  StrategyTypes["LastWeekdayOfMonth"] = "lastWeekdayOfMonth";
  StrategyTypes["NoRepeat"] = "noRepeat";
  StrategyTypes["Stopwatch"] = "stopwatch";
  StrategyTypes["Weekdays"] = "weekdays";
  StrategyTypes["Window"] = "window";
  StrategyTypes["XDayOfXWeekOfMonth"] = "xDayOfxWeekOfMonth";
})(StrategyTypes || (exports.StrategyTypes = StrategyTypes = {}));
