// @azure/core-tracing/dist-esm/implementations/noop/spanNoOpImpl.js
var SpanNoOpImpl = /** @class */ (function () {
    function SpanNoOpImpl() {
    }
    SpanNoOpImpl.prototype.context = function () {
        throw new Error("Method not implemented.");
    };
    SpanNoOpImpl.prototype.setAttribute = function (key, value) {
        throw new Error("Method not implemented.");
    };
    SpanNoOpImpl.prototype.setAttributes = function (attributes) {
        throw new Error("Method not implemented.");
    };
    SpanNoOpImpl.prototype.addEvent = function (name, attributes) {
        throw new Error("Method not implemented.");
    };
    SpanNoOpImpl.prototype.addLink = function (spanContext, attributes) {
        throw new Error("Method not implemented.");
    };
    SpanNoOpImpl.prototype.setStatus = function (status) {
        throw new Error("Method not implemented.");
    };
    SpanNoOpImpl.prototype.updateName = function (name) {
        throw new Error("Method not implemented.");
    };
    SpanNoOpImpl.prototype.start = function (startTime) { };
    SpanNoOpImpl.prototype.end = function (endTime) { };
    SpanNoOpImpl.prototype.isRecordingEvents = function () {
        throw new Error("Method not implemented.");
    };
    return SpanNoOpImpl;
}());
export { SpanNoOpImpl };
