// @azure/core-tracing/dist-esm/implementations/noop/tracerNoOpImpl.js
import { SpanNoOpImpl } from "./spanNoOpImpl";
import { SupportedPlugins } from '../../utils/supportedPlugins';
var TracerNoOpImpl = /** @class */ (function () {
    function TracerNoOpImpl() {
        this.pluginType = SupportedPlugins.NOOP;
    }
    TracerNoOpImpl.prototype.getCurrentSpan = function () {
        throw new Error("Method not implemented.");
    };
    TracerNoOpImpl.prototype.startSpan = function (name, options) {
        return new SpanNoOpImpl();
    };
    TracerNoOpImpl.prototype.withSpan = function (span, fn) {
        throw new Error("Method not implemented.");
    };
    TracerNoOpImpl.prototype.recordSpanData = function (span) {
        throw new Error("Method not implemented.");
    };
    TracerNoOpImpl.prototype.getBinaryFormat = function () {
        throw new Error("Method not implemented.");
    };
    TracerNoOpImpl.prototype.getHttpTextFormat = function () {
        throw new Error("Method not implemented.");
    };
    return TracerNoOpImpl;
}());
export { TracerNoOpImpl };
