// @azure/core-tracing/dist-esm/plugins/noop/noOpSpanPlugin.js
var NoOpSpanPlugin = /** @class */ (function () {
    function NoOpSpanPlugin(span) {
        this._span = span;
    }
    NoOpSpanPlugin.prototype.context = function () {
        return {
            spanId: "",
            traceId: ""
        };
    };
    NoOpSpanPlugin.prototype.end = function (endTime) {
        this._span.end(endTime);
    };
    NoOpSpanPlugin.prototype.start = function (startTime) {
        this._span.start(startTime);
    };
    NoOpSpanPlugin.prototype.setAttribute = function (key, value) {
        throw new Error("Method not implemented.");
    };
    NoOpSpanPlugin.prototype.setAttributes = function (attributes) {
        throw new Error("Method not implemented.");
    };
    NoOpSpanPlugin.prototype.addEvent = function (name, attributes) {
        throw new Error("Method not implemented.");
    };
    NoOpSpanPlugin.prototype.addLink = function (spanContext, attributes) {
        throw new Error("Method not implemented.");
    };
    NoOpSpanPlugin.prototype.setStatus = function (status) {
        throw new Error("Method not implemented.");
    };
    NoOpSpanPlugin.prototype.updateName = function (name) {
        throw new Error("Method not implemented.");
    };
    NoOpSpanPlugin.prototype.isRecordingEvents = function () {
        throw new Error("Method not implemented.");
    };
    return NoOpSpanPlugin;
}());
export { NoOpSpanPlugin };
