// @azure/core-tracing/dist-esm/plugins/noop/noOpTracePlugin.js
import { NoOpSpanPlugin } from "./noOpSpanPlugin";
import { SpanNoOpImpl } from "../../implementations/noop/spanNoOpImpl";
import { SupportedPlugins } from '../../utils/supportedPlugins';
var NoOpTracePlugin = /** @class */ (function () {
    function NoOpTracePlugin(tracer) {
        this.pluginType = SupportedPlugins.NOOP;
        this._tracer = tracer;
    }
    NoOpTracePlugin.prototype.startSpan = function (name, options) {
        var span = new SpanNoOpImpl();
        var noOpSpanPlugin = new NoOpSpanPlugin(span);
        return noOpSpanPlugin;
    };
    NoOpTracePlugin.prototype.getCurrentSpan = function () {
        throw new Error("Method not implemented.");
    };
    NoOpTracePlugin.prototype.withSpan = function (span, fn) {
        throw new Error("Method not implemented.");
    };
    NoOpTracePlugin.prototype.recordSpanData = function (span) {
        throw new Error("Method not implemented.");
    };
    NoOpTracePlugin.prototype.getBinaryFormat = function () {
        throw new Error("Method not implemented.");
    };
    NoOpTracePlugin.prototype.getHttpTextFormat = function () {
        throw new Error("Method not implemented.");
    };
    return NoOpTracePlugin;
}());
export { NoOpTracePlugin };
