// @azure/core-tracing/dist-esm/plugins/opencensus/openCensusTracePlugin.js
import { OpenCensusSpanPlugin } from "../opencensus/openCensusSpanPlugin";
import { SupportedPlugins } from '../../utils/supportedPlugins';
var OpenCensusTracePlugin = /** @class */ (function () {
    function OpenCensusTracePlugin(tracer) {
        this.pluginType = SupportedPlugins.OPENCENSUS;
        this._tracer = tracer;
    }
    OpenCensusTracePlugin.prototype.startSpan = function (name, options) {
        var parent = options
            ? options.parent
                ? options.parent instanceof OpenCensusSpanPlugin
                    ? options.parent.getSpan()
                    : options.parent
                : undefined
            : undefined;
        var span = this._tracer.startChildSpan({
            name: name,
            childOf: parent
        });
        var openCensusSpanPlugin = new OpenCensusSpanPlugin(span);
        return openCensusSpanPlugin;
    };
    OpenCensusTracePlugin.prototype.getCurrentSpan = function () {
        throw new Error("Method not implemented.");
    };
    OpenCensusTracePlugin.prototype.withSpan = function (span, fn) {
        throw new Error("Method not implemented.");
    };
    OpenCensusTracePlugin.prototype.recordSpanData = function (span) {
        throw new Error("Method not implemented.");
    };
    OpenCensusTracePlugin.prototype.getBinaryFormat = function () {
        throw new Error("Method not implemented.");
    };
    OpenCensusTracePlugin.prototype.getHttpTextFormat = function () {
        throw new Error("Method not implemented.");
    };
    return OpenCensusTracePlugin;
}());
export { OpenCensusTracePlugin };
