// @azure/core-tracing/dist-esm/plugins/opencensus/openCensusSpanPlugin.js
import { OpenCensusTraceStatePlugin } from "./openCensusTraceStatePlugin";
var OpenCensusSpanPlugin = /** @class */ (function () {
    function OpenCensusSpanPlugin(span) {
        this._span = span;
    }
    OpenCensusSpanPlugin.prototype.getSpan = function () {
        return this._span;
    };
    OpenCensusSpanPlugin.prototype.end = function (endTime) {
        this._span.end(endTime);
    };
    OpenCensusSpanPlugin.prototype.start = function (startTime) {
        this._span.start(startTime);
    };
    OpenCensusSpanPlugin.prototype.context = function () {
        var openCensusSpanContext = this._span.spanContext;
        return {
            spanId: openCensusSpanContext.spanId,
            traceId: openCensusSpanContext.traceId,
            traceOptions: openCensusSpanContext.options,
            traceState: new OpenCensusTraceStatePlugin(openCensusSpanContext.traceState)
        };
    };
    OpenCensusSpanPlugin.prototype.setAttribute = function (key, value) {
        this._span.addAttribute(key, value);
        return this;
    };
    OpenCensusSpanPlugin.prototype.setAttributes = function (attributes) {
        throw new Error("Method not implemented.");
    };
    OpenCensusSpanPlugin.prototype.addEvent = function (name, attributes) {
        throw new Error("Method not implemented.");
    };
    OpenCensusSpanPlugin.prototype.addLink = function (spanContext, attributes) {
        // Since there is no way to specify the link relationship
        // It is set as Unspecified = 0
        this._span.addLink(spanContext.traceId, spanContext.spanId, 0, attributes);
        return this;
    };
    OpenCensusSpanPlugin.prototype.setStatus = function (status) {
        this._span.setStatus(status.code, status.message);
        return this;
    };
    OpenCensusSpanPlugin.prototype.updateName = function (name) {
        throw new Error("Method not implemented.");
    };
    OpenCensusSpanPlugin.prototype.isRecordingEvents = function () {
        throw new Error("Method not implemented.");
    };
    return OpenCensusSpanPlugin;
}());
export { OpenCensusSpanPlugin };
