// @lighthouse/common/dist/pdf/task/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildTaskPdf = buildTaskPdf;

var _lodash = require("lodash");

var _helpers = require("../helpers");

var _helpers2 = require("../../helpers");

/**
 * buildTaskPdf
 *
 * @param {object} pdfOptions - the pdf options
 * @param {string} pdfOptions.fileTitle - pdf file title
 * @param {function} pdfOptions.footer - function executed to generate footer
 * @param {function} pdfOptions.header - function executed to generate header
 * @param {string} pdfOptions.logoUrl - pdf logo url
 * @param {array} pdfOptions.pageMargins - pdf page margins
 * @param {string} pdfOptions.pageOrientation - pdf page orientation
 * @param {string} pdfOptions.pageSize - pdf page size
 * @param {object} pdfOptions.styles - pdf styles
 * @param {object} pdfOptions.title - pdf title
 * @param {object} data - pdf data
 * @param {object} data.entity - task document
 * @param {object} data.locations - locations documents
 * @param {object} data.settings - settings properties
 * @param {string} data.settings.awsS3BaseUrl - aws S3 base url
 * @param {string} data.settings.cloudinaryBaseUrl - cloudinary base url
 * @param {string} data.timezone - timezone string
 * @param {object} data.users - application user documents
 * @param {object} data.zones - zone documents
 * @returns {Promise} returns pdfmake definition object
 */
function buildTaskPdf(pdfOptions, data) {
  const {
    entity,
    timezone
  } = data;
  const timestamp = entity.createdAt;
  const title = entity.title || 'Unknown';
  const fileTitle = `Task Report - ${title}`;
  return generateContent(data).then(content => (0, _helpers.generateDefinition)({
    content,
    fileTitle,
    timestamp,
    timezone,
    type: 'Task',
    ...pdfOptions
  }));
}

function generateContent(data) {
  const {
    entity
  } = data;
  const {
    entry,
    gps = {},
    title
  } = entity;
  const entityDetails = (0, _helpers2.getTaskEntryDetails)(data);
  const {
    gpsText,
    locationText,
    referenceValue,
    timezoneHourTime
  } = entityDetails;
  const reverseGeocoded = gps.reverseGeocoded;
  const subTitle = `${locationText || gpsText} - ${timezoneHourTime} by ${referenceValue}`;
  const headerSubTitle = (0, _helpers.text)(subTitle, {
    style: 'subTitle'
  });
  const headerTitle = (0, _helpers.text)(title, {
    style: 'title'
  });
  const address = !(0, _lodash.isEmpty)(reverseGeocoded) ? (0, _helpers.getFormattedAddress)(reverseGeocoded) : '';
  const headerAddress = (0, _helpers.text)(address, {
    style: 'small'
  });
  const body = !(0, _lodash.isEmpty)(reverseGeocoded) ? [[headerTitle], [headerSubTitle], [headerAddress]] : [[headerTitle], [headerSubTitle]];
  const titleTable = (0, _helpers.table)({
    body,
    layout: 'noBorders',
    style: 'titleTable'
  });
  return (0, _helpers.buildTemplateContent)(entry.formGroups, data).then(entry => [titleTable, ...entry]);
}
