// @lighthouse/common/lib/helpers/get-audit-entry-details/index.js
import { get } from 'lodash';
import { getAuditItemsData, getLocationReference, getReferenceDetails, getValueAsPercentage, getTimezoneDatetime } from '../';
export function getAuditEntryDetails(data) {
  var entity = data.entity,
      timezone = data.timezone;
  var createdAt = entity.createdAt,
      gps = entity.gps,
      items = entity.items,
      score = entity.score;
  var locationText = getLocationReference(data);
  var gpsText = get(gps, 'reverseGeocoded.label', 'Unknown Location');
  var groupedData = getAuditItemsData(items, data);
  var referenceDetails = getReferenceDetails(data);
  var scoreText = getValueAsPercentage(score.result);
  var timezoneHourTime = getTimezoneDatetime({
    format: 'h:mm a',
    timestamp: createdAt,
    timezone: timezone
  });
  var timezoneDatetime = getTimezoneDatetime({
    timestamp: createdAt,
    timezone: timezone
  });
  return {
    gpsText: gpsText,
    groupedData: groupedData,
    locationText: locationText,
    referenceLabel: referenceDetails.label,
    referenceValue: referenceDetails.value,
    scoreText: scoreText,
    timezoneHourTime: timezoneHourTime,
    timezoneDatetime: timezoneDatetime
  };
}
