// @lighthouse/common/lib/helpers/get-issue-details/index.js
import { get, upperCase } from 'lodash';
import moment from 'moment-timezone';
import { getAssigneesFullNames, getLocationReference, getPrettyDuration, getStatusDetails, getTimezoneDatetime, getUserFullName } from '../';
export function getIssueDetails(data) {
  var entity = data.entity,
      timezone = data.timezone,
      users = data.users;
  var assignees = entity.assignees,
      createdAt = entity.createdAt,
      duration = entity.duration,
      gps = entity.gps,
      status = entity.status,
      user = entity.user;
  var assigneeNames = getAssigneesFullNames(users, assignees);
  var durationInMins = moment.duration(duration, 'minutes');
  var durationText = getPrettyDuration(durationInMins);
  var fullName = getUserFullName(users, user);
  var gpsText = get(gps, 'reverseGeocoded.label', 'Unknown Location');
  var isClosedStatus = status === 'closed';
  var locationText = getLocationReference(data);
  var statusDetails = getStatusDetails(status);
  var statusStyle = statusDetails.style;
  var statusColor = statusStyle.color;
  var statusText = upperCase(statusDetails.text);
  var timezoneDatetime = getTimezoneDatetime({
    timestamp: createdAt,
    timezone: timezone
  });
  return {
    assigneeNames: assigneeNames,
    durationText: durationText,
    gpsText: gpsText,
    fullName: fullName,
    isClosedStatus: isClosedStatus,
    locationText: locationText,
    statusStyle: statusStyle,
    statusColor: statusColor,
    statusText: statusText,
    timezoneDatetime: timezoneDatetime
  };
}
